### PR TITLE
Week5 calendar feature

### PR DESCRIPTION
**Pull request by georgezheng999**
Includes added test features, and is a working implementation of the query as desired in the basic week 5 specification.

This PR is hefty enough by itself, so I've decided to move the optional optimization (which is honestly less of an actual optimization and more-so a different specification with different test cases required all-together) to a following PR.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -2,24 +2,58 @@
...
+    String email = (String) entity.getProperty("creatorEmail");
+    long createdAt = (long) entity.getProperty("createdAt");
+    long parent = (long) entity.getProperty("parent");
+    long root = (long) entity.getProperty("root");
+    this(id, text, createdAt, email, parent, root);
+  }
+
+  //This constructor is associated with the creation of a dummy, root level comment (a supernode for the comment tree)
+  public Comment() {
+    this.id = DUMMY_COMMENT_ID;
```
Comment (qlyao) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: consider calling the other constructor instead of writing all the properties again.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -1,23 +1,67 @@
...
+  // returns a complementary list of open times that are at least as long as the requested duration
+  public Collection<TimeRange> getOpenTimes(List<TimeRange> conflictTimes, MeetingRequest request) {
+    Collections.sort(conflictTimes, TimeRange.ORDER_BY_START);
+    final Collection<TimeRange> queryResult = new ArrayList<>();
+    final int requestDuration = (int) request.getDuration();
+    int maxEndingTimeSoFar = conflictTimes.get(0).end();
+    for (int i = 1; i < conflictTimes.size(); i++) {
+      final int durationDiff = conflictTimes.get(i).start() - maxEndingTimeSoFar; 
+      if (durationDiff >= requestDuration) {
+        queryResult.add(TimeRange.fromStartDuration(conflictTimes.get(i - 1).end(), durationDiff));
```
Comment (qlyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: why is this line needed?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -1,23 +1,67 @@
...
+        queryResult.add(TimeRange.fromStartDuration(maxEndingTimeSoFar, durationDiff));
+      }
+      maxEndingTimeSoFar = max(maxEndingTimeSoFar, conflictTimes.get(i).end());
+      maxEndingTimeSoFar = Math.max(maxEndingTimeSoFar, conflictTimes.get(i).end());
```
Comment (qlyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]: This looks duplicate with the previous line?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -1,23 +1,67 @@
...
+    return false;
+  }
+
+  // Given a list conflictTimes containing times that are not open and a request,
+  // returns a complementary list of open times that are at least as long as the requested duration
+  public Collection<TimeRange> getOpenTimes(List<TimeRange> conflictTimes, MeetingRequest request) {
+    Collections.sort(conflictTimes, TimeRange.ORDER_BY_START);
+    final Collection<TimeRange> queryResult = new ArrayList<>();
+    final int requestDuration = (int) request.getDuration();
+    int maxEndingTimeSoFar = conflictTimes.get(0).end();
```
Comment (qlyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: Is it possible that `conflictTimes` is empty?

I saw you put an empty timeRange `(TimeRange.START_OF_DAY, 0)` before calling this method. If so, I think you can put the empty timeRange inside this function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -1,23 +1,67 @@
...
+      if (eventHasConflict(e, request, false)) {
+        mandConflictTimes.add(e.getWhen());
+      } 
+      if (eventHasConflict(e, request, true)) {
+        mandAndOptConflictTimes.add(e.getWhen());
+      }
+    }
+    final Collection<TimeRange> mandatoryUserTimes = getOpenTimes(mandConflictTimes, request);
+    final Collection<TimeRange> mandAndOptUserTimes = getOpenTimes(mandAndOptConflictTimes, request);
+    return (mandAndOptUserTimes.isEmpty()) ? mandatoryUserTimes : mandAndOptUserTimes;
```
Comment (qlyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: There cannot be two return statements. Does it build?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +273,151 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_C)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
Comment (qlyao) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: give the first param its own line, just like the other two params.

**Comment by georgezheng999** [comment]
integrated changes from pr #10 into this 
also, made sure that the query code passes the test cases -- earlier i couldnt because of security key issues

**Review by qlyao — APPROVED** [review]

**Review by ChuxiangP — APPROVED** [review]
